algebird-core/src/main/scala/com/twitter/algebird/TDigest.scala

```diff
@@ -0,0 +1,257 @@
+}
+
+/** Factory functions for TDigest */
+object TDigest {
+  import scala.language.reflectiveCalls
+  import com.twitter.algebird.Monoid
+
+  /**
+   * Default value for a t-digest delta parameter.  The number of clusters varies, roughly, as
+   * about (50/delta), when data are presented in random order
+   * (it may grow larger if data are not presented randomly).  The default corresponds to
+   * an expected number of clusters of about 100.
+   */
+  val deltaDefault = (50.0 / 100.0) // delta * E[clusters] ~ 50
+
+  /**
+   * The t-digest algorithm will re-cluster itself whenever its number of clusters exceeds
+   * (K/delta).  This value is set such that the threshold is about 10x the heuristically
+   * expected number of clusters for the user-specified delta value.  Generally the number of
+   * clusters will only trigger the corresponding re-clustering threshold when data are being
+   * presented in a non-random order.
+   */
+  val K = 10.0 * 50.0
+
+  /**
+   * Obtain an empty t-digest
+   * @param delta a sketch resolution parameter.
+   * @note Smaller values of delta yield sketches with more clusters, and higher resolution
+   * @note The expected number of clusters will vary (roughly) as (50/delta)
+   */
+  def empty(delta: Double = deltaDefault) = {
+    require(delta > 0.0, s"delta was not > 0")
+    TDigest(delta, 0, TDigestMap.empty)
+  }
+
+  /**
+   * Sketch some data with a t-digest
+   * @param data The data elements to sketch
+   * @param delta The sketch resolution parameter.
+   * @note Smaller values of delta yield sketches with more clusters, and higher resolution
+   * @note The expected number of clusters will vary (roughly) as (50/delta)
+   */
+  def sketch[N](
+    data: TraversableOnce[N],
+    delta: Double = deltaDefault)(implicit num: Numeric[N]) = {
+    require(delta > 0.0, s"delta was not > 0")
+    val td = data.foldLeft(empty(delta))((c, e) => c + ((e, 1)))
+    val ds = scala.util.Random.shuffle(td.clusters.toVector)
+    ds.foldLeft(empty(delta))((c, e) => c + e)
+  }
+
+  /**
+   * Obtain a TDigest monoid instance
+   * @param delta The TDigest sketch resolution parameter
+   * @return a TDigest monoid
+   */
+  def monoid(delta: Double = deltaDefault) = new TDigestMonoid(delta)
+
+  /**
+   * Obtain a TDigest aggregator instance
+   * @tparam N The type of numeric data expected
+   * @param delta The TDigest sketch resolution parameter
+   * @return a TDigest aggregator
+   */
+  def aggregator[N](delta: Double = deltaDefault)(implicit num: Numeric[N]) =
+    new TDigestAggregator[N](delta)
+
+  /** Try to add a bit of efficiency to wasteful construction of t-digest objects for each input */
+  private[algebird] def prepare[N](x: N, delta: Double = deltaDefault)(implicit num: Numeric[N]) =
+    TDigest(delta, 1, TDigestMap.empty + ((num.toDouble(x), 1.0)))
+}
+
+/** The Monoid type class for TDigest objects */
+class TDigestMonoid(delta: Double = TDigest.deltaDefault) extends Monoid[TDigest] {
```

Comment:
I don't like that both TDigest and the monoid know about delta. Can we make it so only one needs to?